plugins/outputs/opentelemetry/client.go

```diff
@@ -0,0 +1,224 @@
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	metricsService "go.opentelemetry.io/proto/otlp/collector/metrics/v1"
+	metricspb "go.opentelemetry.io/proto/otlp/metrics/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// serviceConfig copied from OTel-Go
+	// https://github.com/open-telemetry/opentelemetry-go/blob/a2cecb6e80f6a0712187b080a97f8efb5a61082a/exporters/otlp/internal/otlpconfig/options.go#L47.
+	serviceConfig = `{
+	"methodConfig":[{
+		"name":[
+			{ "service":"opentelemetry.proto.collector.metrics.v1.MetricsService" },
+			{ "service":"opentelemetry.proto.collector.trace.v1.TraceService" }
+		],
+		"retryPolicy":{
+			"MaxAttempts":5,
+			"InitialBackoff":"0.3s",
+			"MaxBackoff":"5s",
+			"BackoffMultiplier":2,
+			"RetryableStatusCodes":[
+				"CANCELLED",
+				"DEADLINE_EXCEEDED",
+				"RESOURCE_EXHAUSTED",
+				"ABORTED",
+				"OUT_OF_RANGE",
+				"UNAVAILABLE",
+				"DATA_LOSS"
+			]
+		}
+	}]
+}`
+)
+
+const maxTimeseriesPerRequest = 500
+
+// client allows reading and writing from/to a remote gRPC endpoint. The
+// implementation may hit a single backend, so the application should create a
+// number of these clients.
+type client struct {
+	logger     telegraf.Logger
+	url        *url.URL
+	timeout    time.Duration
+	tlsConfig  *tls.Config
+	headers    metadata.MD
+	compressor string
+
+	conn *grpc.ClientConn
+}
+
+// connect will dial a new connection if one is not set.  When
+// dialing, this function uses its a new context and the same timeout
+// used for store().
+func (c *client) connect(ctx context.Context) error {
+	if c.conn != nil {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	c.logger.Debugf("new OpenTelemetry connection, url=%s timeout=%s", c.url.String(), c.timeout)
+
+	dopts := []grpc.DialOption{
+		grpc.WithBlock(), // Wait for the connection to be established before using it.
+		grpc.WithDefaultServiceConfig(serviceConfig),
+	}
+	if c.url.Scheme != "http" {
+		dopts = append(dopts, grpc.WithTransportCredentials(credentials.NewTLS(c.tlsConfig)))
+	} else {
+		dopts = append(dopts, grpc.WithInsecure())
+	}
+	if c.compressor != "" && c.compressor != "none" {
+		dopts = append(dopts, grpc.WithDefaultCallOptions(grpc.UseCompressor(c.compressor)))
+	}
+	address := c.url.Hostname()
+	if len(c.url.Port()) > 0 {
+		address = net.JoinHostPort(address, c.url.Port())
+	}
+	conn, err := grpc.DialContext(ctx, address, dopts...)
+	if err != nil {
+		return err
+	}
+	c.conn = conn
+	return err
+}
+
+// ping sends an empty request the endpoint.
+func (c *client) ping(ctx context.Context) error {
+	// Loop until the context is canceled, allowing for retryable failures.
+	for {
+		err := c.connect(ctx)
```

Comment:
To be honest, this is somewhat unexpected from my perspective. The task of this function is to test connectivity, so I'm not sure if it's a good idea to connect. However, if everybody else accepts this as is, I can live with it.